File: trabalho_cpp/marcos_rogerio/src/Cliente.cpp
#include <map>
#include <set>
#include <fstream>
#include <iostream>
#include <string>
#include "Cliente.h"
#include "util/Tokenizer.h"
#include "util/DateUtils.h"
#include "util/NumberUtils.h"

using namespace std;
using namespace br_ufes_inf_nemo_cpp_util;

namespace trabalho_cpp {

//Construtor 
Cliente::Cliente(int identificador, string nome, string endereco,
		string telefone, time_t data) {
	this->identificador = identificador;
	this->nome = nome;
	this->endereco = endereco;
	this->telefone = telefone;
	this->data = data;
}

//Método que será implementado nas subclasses para polimorfismo
string Cliente::lucro(map<int, Venda> mapa_v, map<int, Produto> mapa_p) {
	return "";
}

Cliente::~Cliente() {}

//Construtor da subclasse PessoaFisica
PessoaFisica::PessoaFisica(int identificador, string nome, string endereco,
		string telefone, time_t data, string cpf) :
	Cliente(identificador, nome, endereco, telefone, data) {
	this->cpf = cpf;
}

//Função que retorna uma linha a ser impressa no arquivo "2-areceber.csv"
string PessoaFisica::lucro(map<int, Venda> mapa_v, map<int, Produto> mapa_p) {
	double lucro_d = 0;

	//Calcula o lucro e receita para cada PessoaFisica
	for (auto& v : mapa_v) {
		if(this->identificador == (v.second).num_cliente())
		lucro_d = lucro_d + (v.second).receita(mapa_p);
	}

	return (nome + ";F;" + cpf + ";" + telefone + ";" + formatDate(data,
			DATE_FORMAT_PT_BR_SHORT) + ";R$ " + formatDoubleCurrency(lucro_d,
			LOCALE_PT_BR));

}

PessoaFisica::~PessoaFisica() {}


//Construtor da subclasse PessoaJuridica
PessoaJuridica::PessoaJuridica(int identificador, string nome, string endereco,
		string telefone, time_t data, string cnpj, int inscricao) :
	Cliente(identificador, nome, endereco, telefone, data) {
	this->cnpj = cnpj;
	this->inscricao = inscricao;
}

PessoaJuridica::~PessoaJuridica() {}

//Função que retorna uma linha a ser impressa no arquivo "2-areceber.csv"
string PessoaJuridica::lucro(map<int, Venda> mapa_v, map<int, Produto> mapa_p) {
	double lucro_d = 0;

	//Calcula o lucro e receita para cada PessoaJuridica
	for (auto& v : mapa_v) {
		if(this->identificador == (v.second).num_cliente())
		lucro_d = lucro_d + (v.second).receita(mapa_p);
	}

	return (nome + ";J;" + cnpj + ";" + telefone + ";" + formatDate(data,
			DATE_FORMAT_PT_BR_SHORT) + ";R$ " + formatDoubleCurrency(lucro_d,
			LOCALE_PT_BR));
}


}

File: trabalho_java/Marcos_Rogerio/src/marcos_rogerio/pessoas/Cliente.java
package marcos_rogerio.pessoas;

import java.io.*;
import java.text.DateFormat;
import java.util.*;

import marcos_rogerio.produtos.Venda;

//Classe abstrata que representa uma pessoa qualquer

public abstract class Cliente implements Serializable{

	protected int identificador;
	protected String nome;
	protected String endereco;
	protected String telefone;
	protected Date dataInicio;

	// Construtor da classe abstrata

	public Cliente(int identificador, String nome, String endereco,
			String telefone, Date dataInicio) {

		this.identificador = identificador;
		this.nome = nome;
		this.endereco = endereco;
		this.telefone = telefone;
		this.dataInicio = dataInicio;

	}

	// Sobrescrita do método toString() da classe Object

	@Override
	public String toString() {
		return this.nome;
	}

	// Método que adciona clientes a um Map de clientes, a partir da leitura de
	// um arquivo

	public static void addClientes(Map<Integer, Cliente> mapa, File arquivo) {

		try (Scanner scanner = new Scanner(new BufferedReader(new FileReader(
				arquivo)))) {

			scanner.useDelimiter(";|\\n");
			scanner.useLocale(new Locale("pt", "BR"));

			// Pula a linha das descricoes do arquivo
			scanner.nextLine();

			while (scanner.hasNext()) {
				int identificador = scanner.nextInt();
				String nome = scanner.next();
				String endereco = scanner.next();
				String telefone = scanner.next();

				// formatando e armazenando a data
				DateFormat df = DateFormat.getDateInstance(DateFormat.MEDIUM);
				Date data = df.parse(scanner.next());

				// verificando se a pessoa e fisica ou juridica e armazenando o
				// cliente
				switch (scanner.next()) {

				case "F":
					String cpf = scanner.next();
					mapa.put(identificador, new PessoaFisica(identificador,
							nome, endereco, telefone, data, cpf));

					// descarta o ponto e virgula excedente
					if (scanner.hasNext())
						scanner.next();
					break;
				case "J":
					String cnpj = scanner.next();
					int inscricao = scanner.nextInt();
					mapa.put(identificador, new PessoaJuridica(identificador,
							nome, endereco, telefone, data, cnpj, inscricao));
					break;
				default:
					throw new IOException();

				}

			}

			scanner.close();
		} catch (Exception e) {
			System.out.println("Erro de I/O.");

		}

	}

	// Método abstrato que será implementado pelas subclasses, para uso do
	// polimorfismo

	public abstract String lucro(Set<Venda> conjunto);

	// Método que retorna o identificador do Cliente

	public int identificador() {
		return this.identificador;
	}

	// Método que escreve o arquivo "2-areceber"

	public static void areceber(Map<Integer, Cliente> mapa,
			Set<Venda> conjunto, File arquivo) {

		String[] linhas = new String[mapa.size()];
		int ind = 0;

		for (Integer i : mapa.keySet()) {
			linhas[ind] = mapa.get(i).lucro(conjunto);
			ind++;
		}

		Arrays.sort(linhas);

		try {
			PrintWriter writer = new PrintWriter(arquivo);

			if (!arquivo.exists()) {
				arquivo.createNewFile();
			}

			// imprime o cabeçalho
			writer.println("Cliente;Tipo;CPF/CNPJ;Telefone;Data de cadastro;Total a pagar");

			for (int i = 0; i < mapa.size(); i++) {
				writer.println(linhas[i]);
			}

			writer.close();
		} catch (IOException e) {
			System.out.println("Erro de I/O.");
		}

	}
}
